### PR TITLE
feat(server): add bottle server module

### PR DESCRIPTION
**Pull request by ziadsawalha**
Includes code for eventlet and tornado adapters
with customizations.

xtornado: enables xheaders always.
xeventlet: supports SSL for testing and additional
    tuning parameters.

### CODE REVIEW

File: simpl/server.py [hunk 1]
```diff
@@ -0,0 +1,158 @@
...
+
+import logging
+import os
+
+import bottle
+
+LOG = logging.getLogger(__name__)
+
+
+class EventletLogFilter(object):  # pylint: disable=R0903
```
Comment (larsbutler) [inline simpl/server.py]: Why do we have `pylint: disable` statements in `simpl`? We're not even using `pylint`, as far as I can tell (despite the fact that we have a pylintrc file).

Comment (ziadsawalha) [inline simpl/server.py]: I prefer to be explicit. Especially when doing anything that can be flagged by one of the common linters.

Comment (larsbutler) [inline simpl/server.py]: But it doesn't actually do anything. Why don't start using pylint then? Otherwise, we're just adding superfluous comments to the codebase.

File: simpl/server.py [hunk 2]
```diff
@@ -0,0 +1,158 @@
...
+    def __init__(self, conf, log, access_log=None):
+        """Initialize with config and optional access_log."""
+        self.config = conf
+        self.log = log
+        self.access_log = access_log
+
+    def write(self, text):
+        """Write to appropriate target."""
+        if text:
+            if text[0] in '(w':
```
Comment (larsbutler) [inline simpl/server.py]: What is the significance of `(w`?

**Comment by larsbutler** [comment]
`server.py` has 23% test coverage.

```
Name                    Stmts   Miss  Cover   Missing
-----------------------------------------------------
simpl                       5      0   100%   
simpl.__about__            11      0   100%   
simpl.config              347     43    88%   259, 264-265, 330-333, 339, 376-378, 390, 396, 425, 429, 433, 437, 444, 487, 510-511, 568, 593-597, 632, 636-637, 648, 661, 811-816, 835-842, 860-864, 867
simpl.exceptions           35      8    77%   42, 72, 78-83, 108
simpl.git                 299     10    97%   139, 199-200, 281, 333, 431, 541, 729-730, 732
simpl.incubator             1      0   100%   
simpl.incubator.dicts     147      7    95%   49, 68, 147-148, 155, 248, 261
simpl.log                  49     32    35%   87-94, 104-111, 124-133, 140-153, 165, 169, 174, 177
simpl.middleware            0      0   100%   
simpl.middleware.cors      43      3    93%   34-35, 106
simpl.rest                 82      3    96%   39, 54, 192
simpl.secrets              12      5    58%   32-36
simpl.server               70     54    23%   38-40, 44-51, 83-109, 113-135, 139, 150-156
simpl.threadlocal          35      2    94%   84, 113
simpl.utils                 0      0   100%   
simpl.utils.shell          26      1    96%   69
-----------------------------------------------------
TOTAL                    1162    168    86%   
----------------------------------------------------------------------

```

File: simpl/server.py [hunk 3]
```diff
@@ -0,0 +1,158 @@
...
+                    'ssl_version', 'ca_certs', 'do_handshake_on_connect',
+                    'suppress_ragged_eofs', 'ciphers'):
+            try:
+                ssl_args[arg] = self.options.pop(arg)
+            except KeyError:
+                pass
+        address = (self.host, self.port)
+        try:
+            sock = eventlet.listen(address, **socket_args)
+        except TypeError:
```
Comment (larsbutler) [inline simpl/server.py]: What causes this to throw a `TypeError`?

File: simpl/server.py (hunk 1 above)
Comment (ziadsawalha) [inline simpl/server.py]: You raise a valid question which we should bring up for the team to discuss.  Most of our other projects run pylint in the tox runs. I would vote to include it (with a customized pylintrc file). I've raised #55 to host that discussion.

WRT to this PR, pylint comments are in many other places in this codebase (as well as in many of our other projects - most of which actually include pylint in the tox runs) - as well as a pylintrc file - so this line does not diverge from what we've been doing.

It does document that we've intentionally not included public methods in this class. And it does  it in a way that is compatible with one of the most commonly used linting tools for python.

While removing this line is simple, I'd rather stay consistent with the rest of the codebase and remove it if we choose to not use pylint and remove pylint comments from the whole code base.

File: simpl/server.py (hunk 2 above)
Comment (ziadsawalha) [inline simpl/server.py]: It's what eventlet preceds its WSGI event messages with; we write those as debug messages. The other messages are apache log-style API messages.

Comment (ziadsawalha) [inline simpl/server.py]: This is ported from https://github.com/bottlepy/bottle/blob/master/bottle.py#L2997. Based on the comment I would say some version of eventlet.

File: simpl/server.py (hunk 3 above)
Comment (ziadsawalha) [inline simpl/server.py]: This is ported from https://github.com/bottlepy/bottle/blob/master/bottle.py#L2997. Based on the comment I would say some version of eventlet.

**Comment by ziadsawalha** [comment]
@larsbutler WRT test coverage, I started down the route of copying over tests from https://github.com/bottlepy/bottle. Before I go down that route, is that an acceptable approach?

FWIW, it calls out to subprocesses to accomplish the testing so it won't impact coverage stats. I don't know if test coverage is the metric we want to strive for in this case.

File: simpl/server.py (hunk 2 above)
Comment (larsbutler) [inline simpl/server.py]: Ah, I never would have guessed that. Is this documented somewhere?

File: simpl/server.py [hunk 4]
```diff
@@ -0,0 +1,158 @@
...
+    """Receives eventlet log.write() calls and routes them.
+
+    Thread and WSGI messages only get written in debug mode.
+    API calls get written as INFO and to an access_log if provided.
+
+    An instance of this class can be passed in to a bottle.run command using
+    the `log` keyword.
+    """
+
+    def __init__(self, conf, log, access_log=None):
```
Comment (larsbutler) [inline simpl/server.py]: Please add `param/keyword` docs for these parameters. I could make a guess at what `log` is, but it's not clear what the others are.

Also, `conf/self.config` doesn't appear to be used anywhere.

File: simpl/server.py [hunk 5]
```diff
@@ -0,0 +1,158 @@
...
+        self.config = conf
+        self.log = log
+        self.access_log = access_log
+
+    def write(self, text):
+        """Write to appropriate target."""
+        if text:
+            if text[0] in '(w':
+                # write thread and wsgi messages to debug only
+                self.log.debug(text[:-1])
```
Comment (larsbutler) [inline simpl/server.py]: What are we chopping off here? What's is the expected value of the last item?

File: simpl/server.py [hunk 6]
```diff
@@ -0,0 +1,158 @@
...
+    the `log` keyword.
+    """
+
+    def __init__(self, conf, log, access_log=None):
+        """Initialize with config and optional access_log."""
+        self.config = conf
+        self.log = log
+        self.access_log = access_log
+
+    def write(self, text):
```
Comment (larsbutler) [inline simpl/server.py]: What is `text` exactly? I would expect it to be a string, but below it's treated like a list. There's also no test code which exercises it, so I have no way of knowing a) what the method expects; b) how I'm supposed to use the class; or c) whether or not it is implemented correctly.

File: simpl/server.py (hunk 5 above)
Comment (ziadsawalha) [inline simpl/server.py]: Added docs and tests

File: simpl/server.py (hunk 2 above)
Comment (ziadsawalha) [inline simpl/server.py]: No, but I added the docs now.

File: simpl/server.py (hunk 5 above)
Comment (ziadsawalha) [inline simpl/server.py]: We're chopping off the carriage return

File: all-requirements.txt [hunk 7]
```diff
@@ -3,5 +3,9 @@
 # For simpl.rest
 bottle==0.12.8
 
+# For simpl.server
+eventlet==0.17.4
+tornado==4.2
```
Comment (Pablosan) [inline all-requirements.txt]: I know the pattern predates this PR, but would it make more sense for module-specific requirements to be listed in a module-specific requirements file?

Comment (ziadsawalha) [inline all-requirements.txt]: May I submit another PR for that after this?

**Comment by ziadsawalha** [comment]
@larsbutler I added integration tests for the Tornado and Eventlet server adapters. Because they fork, they don't show up as covered. I also added tests for the Eventlet log writer.

File: all-requirements.txt (hunk 7 above)
Comment (Pablosan) [inline all-requirements.txt]: Yes, absolutely. I'm not even sure separate files is the right approach. We could use a follow-up PR to facilitate that discussion.
